src/main/java/io/confluent/idesidecar/restapi/clients/SchemaErrors.java

```diff
@@ -0,0 +1,85 @@
+package io.confluent.idesidecar.restapi.clients;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@ApplicationScoped
+public class SchemaErrors {
+
+  private final Map<String, Map<String, String>> cacheOfCaches = new ConcurrentHashMap<>();
+
+  private String generateKey(String connectionId, String clusterId) {
+    return connectionId + ":" + clusterId;
+  }
+
+  private String generateSchemaKey(int schemaId) {
+    return "schema:" + schemaId;
+  }
+
+  public Map<String, String> getSubCache(String key) {
+    return cacheOfCaches.computeIfAbsent(key, k -> new ConcurrentHashMap<>());
+  }
+
+  // Methods for connection and cluster ID
+  public String readByConnectionId(String connectionId, String clusterId) {
+    return getSubCache(generateKey(connectionId, clusterId)).get(clusterId);
+  }
+
+  public void writeByConnectionId(String connectionId, String clusterId, String error) {
+    getSubCache(generateKey(connectionId, clusterId)).put(clusterId, error);
+  }
+
+  public void clearByConnectionId(String connectionId, String clusterId) {
+    Map<String, String> subCache = getSubCache(generateKey(connectionId, clusterId));
+    subCache.remove(clusterId);
+    if (subCache.isEmpty()) {
+      cacheOfCaches.remove(generateKey(connectionId, clusterId));
+    }
+  }
```

Comment:
I wonder if we should have a method that response to connection events, like on `ClusterCache`, when the connection is closed:
```
  void onConnectionDeleted(@ObservesAsync @Lifecycle.Deleted ConnectionState connection) {
    errorsBySchemaId.invalidate(connection.getSpec().id());
  }
```
The risk is that we remove the errors when a connection is closed, but meanwhile the sidecar is still processing a previously-submitted consume request that still needs the schema failures.

On second thought, we could not have this method and instead rely upon the CONNECTION_TTL to automatically evict from the `errorsByConnectionId` cache the `ErrorsForConnection` entry for a disconnected/closed connection, some TTL after it is unused. I suspect the sizes of these caches will not be huge (home many failed fetches by schema ID might one connection have), so it's okay to keep them around for a bit longer than the connection exists.

The actual TTL depends a lot on the SCHEMA_FETCH_ERROR_TTL, which corresponds to the existing `RecordDeserializer.CACHE_FAILED_SCHEMA_ID_FETCH_DURATION` that is only 30 seconds. Basically, after 30 seconds of _non-use_, the error for a `SchemaId` will be purged anyway, which means after 30 seconds, the `ErrorsForConnection` will be empty anyway. So if `CACHE_FAILED_SCHEMA_ID_FETCH_DURATION` is 30 seconds, then maybe `CONNECTION_TTL` could be 60 seconds (since a thread processing the consume request my hold onto the `ErrorsForConnection` object for 0-5-ish seconds, with a bit of a buffer).

But, we probably **do** want to clear any `ErrorsForConnection` entry when a connection is created or updated, just in case the SR URI is changed on the connection:
```
  void onConnectionCreated(@ObservesAsync @Lifecycle.Created @Lifecycle.Updated ConnectionState connection) {
    errorsBySchemaId.invalidate(connection.getSpec().id());
  }
```

Reply:
> The risk is that we remove the errors when a connection is closed, but meanwhile the sidecar is still processing a previously-submitted consume request that still needs the schema failures.

I'm not sure if there's a real risk. In this case, the sidecar (or the used SR client) would hit the proxied SR routes again, which will return a proxy error because the connection ID does no longer exist. So, invalidating the cache after deleting a connection could make sense because the errors will change.

> On second thought, we could not have this method and instead rely upon the CONNECTION_TTL to automatically evict from the errorsByConnectionId cache the ErrorsForConnection entry for a disconnected/closed connection, some TTL after it is unused. I suspect the sizes of these caches will not be huge (home many failed fetches by schema ID might one connection have), so it's okay to keep them around for a bit longer than the connection exists.

Good point. From my POV, cache invalidation is less about saving space but more about returning correct responses. Let's assume a user runs into errors because they have accidentally provided incorrect SR credentials. If we don't invalidate the cache when they fix the credentials by recreating or updating the connection, we'd continue serving the errors despite valid credentials being present.

> But, we probably do want to clear any ErrorsForConnection entry when a connection is created or updated

Yeah - I think, for the sake of keeping the implementation as simple and correct as possible, we should continue using the `SCHEMA_FETCH_ERROR_TTL`, not introduce the `CONNECTION_TTL`, and invalidate the cache when a connection has been created/updated/deleted. WDYT?